### ISSUE TITLE
Add custom name support to `newrelic.ExternalSegment`

### ISSUE THREAD

**Issue by justiandre** [newrelic/go-agent#87]
# `newrelic.ExternalSegment` with support for name

We would like to be able to enter the name in the `newrelic.ExternalSegment` to solve the problem of calls to `APIGateway`/`Router`.

In `MicroServices` the `APIGateway` is widely used to centralize calls to others API, hence, external calls are centralized at the `APIGateway`.
This makes analyzing different paths to the same host difficult from newrelic's main dashboard. 
This could be easily solved if the `newrelic.ExternalSegment` supported a name.

<img width="546" alt="Captura de Tela 2019-05-03 às 10 13 40" src="https://user-images.githubusercontent.com/1108040/57167897-d7f30380-6dd5-11e9-8553-e04d40a0efe3.png">

As shown in the drawing, the `newrelic.ExternalSegment` only registers the call to the host` api_gateway` even though it is the same host, I would like to register the calls differently,  calls to `api_gateway/service_01`  a segment  and calls to`api_gateway/service_02` other segment.

**How it is registered**
<img width="568" alt="Captura de Tela 2019-05-04 às 09 29 08" src="https://user-images.githubusercontent.com/1108040/57179149-d4509280-6e50-11e9-9c8d-d4997108071b.png">

**How I would like register**
<img width="566" alt="Captura de Tela 2019-05-04 às 09 33 06" src="https://user-images.githubusercontent.com/1108040/57179148-d4509280-6e50-11e9-9228-4759b199e27a.png">

--------

The implementation could follow the following pattern: If the name of `newrelic.ExternalSegment` is entered it registers this name, if it is not informed it registers the Host (current behavior), in this way it would not change the behavior of the applications that already use it.

--------

This would be a great evolution, for the applications we build here on @mercadolibre.
Can I follow through with this implementation? Is this change feasible?

**Comment by willnewrelic** [comment]
Hi @justiandre 

Thanks for writing in!  Host level external metrics do not have the specificity you need:  You want to be able to add the first segment of the path to the metric, or control the name of the metric directly.  I see how that would be useful!

I have brought this idea to the attention of our product management.

For the time being, I recommend using the `URL` field of the `ExternalSegment`:  You can create a fictitious url with the first path segment in the hostname.  Example:

```go
	s := newrelic.ExternalSegment{
		StartTime: newrelic.StartSegmentNow(txn),
		URL:       "http://api_gateway_service_02",
	}
	// external call here
	s.End()
```

Just be careful to ensure that `URL` contains a scheme:  It will be parsed by the standard library's `url.Parse`.

**Comment by willnewrelic** [comment]
Hey @justiandre 

Here is some code that demonstrates the interim solution described above:  This code creates external metrics `External/example.com_zip/all` from requests to `http://example.com/zip/zap`.

```go
func startExternal(txn newrelic.Transaction, req *http.Request) *newrelic.ExternalSegment {
	segment := newrelic.StartExternalSegment(txn, req)
	if u := req.URL; nil != u {
		var firstSegment string
		if s := strings.Split(strings.TrimPrefix(u.Path, "/"), "/"); len(s) > 0 {
			firstSegment = s[0]
		}
		nu := url.URL{
			Scheme: u.Scheme,
			Host:   u.Host + "_" + firstSegment,
		}
		segment.URL = nu.String()
	}
	return segment
}

func doRequest(txn newrelic.Transaction) {
	client := &http.Client{}
	request, _ := http.NewRequest("GET", "http://example.com/zip/zap", nil)
	segment := startExternal(txn, request)
	defer segment.End()
	resp, err := client.Do(request)
	if nil != err {
		return
	}
	defer resp.Body.Close()
	fmt.Println(resp.StatusCode)
}
```

We acknowledge that this is bothersome, and we are discussing improvements and additional ExternalSegment fields!

**Comment by justiandre** [comment]
Hello, how's the progress of this issue?

**Comment by justiandre** [comment]
> Ei @justiandre
> 
> Veja alguns códigos que demonstram a solução provisória descrita acima: Esse código cria métricas externas `External/example.com_zip/all`de solicitações para `http://example.com/zip/zap`.
> 
> ```go
> func  startExternal ( txn  newrelic . Transação , req * http . Request ) * newrelic . ExternalSegment {
> 	 segment  : = newrelic. StartExternalSegment (txn, req)
> 	 se  u  : = req. URL ; nil ! = u {
> 		 var  firstSegment  string 
> 		se  s  : = strings. Dividir (strings. TrimPrefix (u. Caminho , " / "), " / " ); len (s)> 0 {
> 			firstSegment = s [ 0 ]
> 		}
> 		nu  : = url. URL {
> 			Esquema: u. Esquema ,
> 			Anfitrião: u. Host + " _ " + firstSegment,
> 		}
> 		segmento. URL = nu String ()
> 	}
> 	segmento de retorno
> }
> 
> func  doRequest ( txn  newrelic . Transação ) {
> 	 client  : = & http. Pedido do cliente {}
> 	 , _ : = http. NewRequest ( " GET " , " http://example.com/zip/zap " , nil )
> 	 segmento : = startExternal (txn, pedido)
> 	 adiar segmento. End ()
> 	 resp , err : = client. Faça (pedido)
> 	 se nulo     ! = err {
> 		 return
> 	}
> 	adiar resp. Corpo . Fechar ()
> 	fmt. Println (resp. StatusCode )
> }
> ```
> 
> Reconhecemos que isso é incômodo e estamos discutindo melhorias e campos adicionais do ExternalSegment!

Today we are doing so, but it is a very very inelegant solution and that does not make sense !! It is difficult to generalize it into other components.

**Comment by justiandre** [comment]
> Hey @justiandre
> 
> Here is some code that demonstrates the interim solution described above: This code creates external metrics `External/example.com_zip/all` from requests to `http://example.com/zip/zap`.
> 
> ```go
> func startExternal(txn newrelic.Transaction, req *http.Request) *newrelic.ExternalSegment {
> 	segment := newrelic.StartExternalSegment(txn, req)
> 	if u := req.URL; nil != u {
> 		var firstSegment string
> 		if s := strings.Split(strings.TrimPrefix(u.Path, "/"), "/"); len(s) > 0 {
> 			firstSegment = s[0]
> 		}
> 		nu := url.URL{
> 			Scheme: u.Scheme,
> 			Host:   u.Host + "_" + firstSegment,
> 		}
> 		segment.URL = nu.String()
> 	}
> 	return segment
> }
> 
> func doRequest(txn newrelic.Transaction) {
> 	client := &http.Client{}
> 	request, _ := http.NewRequest("GET", "http://example.com/zip/zap", nil)
> 	segment := startExternal(txn, request)
> 	defer segment.End()
> 	resp, err := client.Do(request)
> 	if nil != err {
> 		return
> 	}
> 	defer resp.Body.Close()
> 	fmt.Println(resp.StatusCode)
> }
> ```
> 
> We acknowledge that this is bothersome, and we are discussing improvements and additional ExternalSegment fields!

Today we are doing so, but it is a very very inelegant solution and that does not make sense !! It is difficult to generalize it into other components.

**Comment by willnewrelic** [comment]
Hi @justiandre

We just released version 2.9.0 which includes changes you might find useful!

`ExternalSegment` now has three new optional fields: `Host`, `Procedure`, and `Library`.

* `Host` is used for external metrics, transaction trace segment names, and span event names.
* `Procedure` is used for transaction breakdown metrics.
* `Library` is used for external metrics and the "component" span attribute.  "http" is the default.

The easier thing to do is just populate `Host` field with the name you would like to see:

```
    segment := newrelic.ExternalSegment{
        StartTime: txn.StartSegmentNow(),
        Host:      "api_gateway_service_01",  // slashes are not allowed, so use underscores
    }
```

Alternatively, you can populate both `Host` and `Procedure`.  Unfortunately, the External tab metrics do not include `Procedure`:  It will only appear in the metrics in each transaction breakdown.

```
    segment := newrelic.ExternalSegment{
        StartTime: txn.StartSegmentNow(),
        Host:      "api_gateway",
        Procedure: "service_01",
    }
```

We hope these new fields make things better for you!

**Comment by willnewrelic** [comment]
Hi @justiandre !

I hope the new `ExternalSegment` fields are working for you!  I'm going to close this:  Don't hesitate to re-open this or reach out if you would like to discuss further!
